### PR TITLE
[FDS-2386] Synapse entity tracking and code concurrency updates

### PR DESCRIPTION
**Pull request by BryanFauble**
I will be adding comments to specific areas of the code that should be looked at and carefully considered. Below covers the overall gist of what is going on here:

**Problem:**

1. When an entity is retrieved from the synapse python client it is not stored for further use within the context of a request hitting schematic. The entity is re-retrieved several times throughout the lifecycle of the request causing significant bottlenecks when retrieving data out of Synapse.
2. When downloading manifests a temporary directory is being used when running on AWS, but not locally, or for integration tests. As a result concurrent requests in those contexts were causing manifest files to be overwritten for new tests and causing previous tests to fail.
3. A few concepts for great expectations (suites and checkpoints) were constantly being created and torn down, even during requests where the checkpoint/suite was not created. This leads to more concurrency type issues where a checkpoint/suite could be torn down in the middle of a request, or at a point in time it isn't expected.

**Solution:**

1. Creating a simple "pull through cache" that can be added via composition to the classes that interact with Synapse. This is acting as a small (albeit hacky) shim layer between schematic and SYNPY. The usage of this entity tracker allows the code to function largely as it has been without the need to more in-depth refactoring. This means that an entity for a user session is pulled out of memory before it is pulled from Synapse again.
2. Writing manifests to a temporary directory in all cases except for when being used as a CLI tool.
3. Updating the logic for tearing down the great expectation suite and checkpoints to only consider tearing down the checkpoints that itself created.

**Testing:**

1. I have been stepping through the code with the integration tests that have been written to verify functionality.
2. Will need to have a verbose testing plan in place to verify that schematic continues to function as expected. @linglp @andrewelamb Are you two available to help with the concepts of a plan for helping to test all of these changes?

### CODE REVIEW

File: pyproject.toml
Comment (BryanFauble) [inline pyproject.toml:59]: I have a few fixes I've found and added for SYNPY that will need to be included. For now I am pointing to the GIT branch that has the changes, but will plan to release this before I merged this into develop.

File: pyproject.toml
Comment (BryanFauble) [inline pyproject.toml:79]: flask/jaeger are deprecated. Swapping over to the otel-api/sdk libraries instead.

File: pyproject.toml
Comment (BryanFauble) [inline pyproject.toml:138]: These were moved to `pytest.ini` to stop a bunch of warning from being emitted.

File: schematic/__init__.py
Comment (BryanFauble) [inline schematic/__init__.py:59]: Most all the code in this script was just moved except for this `FlaskInstruement` and the request/response hook. This is coming from this library: https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-flask - Which allows auto instrumentation of the flask server that is running here

File: schematic/manifest/commands.py
Comment (BryanFauble) [inline schematic/manifest/commands.py:350]: When running in CLI mode we will not be downloading manifest files to a temporary folder, instead continue to use the defined manifest folder from the config.

File: schematic/models/GE_Helpers.py
Comment (BryanFauble) [inline schematic/models/GE_Helpers.py:120]: This removes extra calls from leaving schematic.

File: schematic/models/GE_Helpers.py
Comment (BryanFauble) [inline schematic/models/GE_Helpers.py:158]: I found issues when tests were running concurrently where great expectation test suites were stepping over eachother, causing tests to break in Un-expected places. With this change the JSON file that is created for each test suite run is unique, and cleaned up after the test suite has ran.

No more will another test clean these up.

Comment (BryanFauble) [reply schematic/models/GE_Helpers.py:158]: The same notes about concurrency also apply when this code is running as an API server. If 2 users were sending in a request at the same time this is likely to have caused issues if they lined up just perfectly.

File: schematic/models/GE_Helpers.py
Comment (BryanFauble) [inline schematic/models/GE_Helpers.py:412]: Using a unique checkpoint name to prevent name collisions with other GE runs.

File: schematic/models/GE_Helpers.py
Comment (BryanFauble) [inline schematic/models/GE_Helpers.py:365]: @andrewelamb @linglp @milen-sage - Is this code needed? It builds a bunch of HTML files, but I couldn't understand if they were actually used anywhere?

File: schematic/models/metadata.py
Comment (BryanFauble) [inline schematic/models/metadata.py:279]: No longer cleaning up the test suite run here. Instead relying on the cleanup to happen in the `finally` block after the expectation runs.

File: schematic/models/validate_manifest.py
Comment (BryanFauble) [inline schematic/models/validate_manifest.py:164]: Some extra code to help guard against extra things from being tracked in memory (It's super minor)

File: schematic/models/validate_manifest.py
Comment (BryanFauble) [inline schematic/models/validate_manifest.py:185]: Unique identifier to prevent name collisions

File: schematic/models/validate_manifest.py
Comment (BryanFauble) [inline schematic/models/validate_manifest.py:332]: This is just automatically re-formatted code. I did not make any changes here.

File: schematic/store/synapse_tracker.py
Comment (BryanFauble) [inline schematic/store/synapse_tracker.py:1]: This file is the main star of all these changes. Please read carefully.

File: schematic/store/synapse.py
Comment (BryanFauble) [inline schematic/store/synapse.py:126]: Instead of blanket removing previously downloaded manifest files instead this will have a 1 hour delay to clean them up. This was implemented to prevent a manifest from getting cleaned up that might be actively being used, or used for a successful cache pull (ie: Not downloading the manifest file again).

File: schematic/store/synapse.py
Comment (BryanFauble) [inline schematic/store/synapse.py:175]: This is a potentially impactful change here @andrewelamb @linglp @milen-sage that could use a good look. Basically what this is solving:

(When NOT running on AWS)
1. When manifest were downloaded the `downloadAs` name, or the name of the actual file, could conflict with a file that was already in the location being downloaded to. This would overwrite the file and potentially cause the test, or whatever kick off that code to fail.

When running on either AWS or not

1. When downloading an entity where we already have a local copy of it (Think of a file that exists within the synapse cachemap) it will copy that file over to the requested location. The issue however is that the file that is copied retains whatever name it has been renamed to (Some tests do this, and code in schematic does this). Because of this I have the `os.rename` logic in place to set the name of the copied file back to the expected file name so that any subsequent code behaves as expected.

File: schematic/store/synapse.py
Comment (BryanFauble) [inline schematic/store/synapse.py:588]: There are a few things to note here on these changes:

1. I have added a change to SYNPY (https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1130) that stored the ID of the user profile that logged into Synapse. This removes the need to call to `getUserProfile()` and saves an extra HTTP call.
2. I am storing the result of getting the paginated list of project headers into the `synapse_entity_tracker`. It is really slow to crawl this information and it occurs several times. By doing this the content is re-used in the context of the request.
3. I removed the need to call `syn.get()` for each of the projects as the paginated list of project headers contained all the data we needed.

File: schematic/store/synapse.py
Comment (BryanFauble) [inline schematic/store/synapse.py:1499]: Swapping to `findEntityId` prevents the need to walk the project to find the applicable table ID. Instead we can utilize the Synapse Rest API to find the ID for us.

Comment (BryanFauble) [reply schematic/store/synapse.py:1499]: This will return `None` if not found

File: schematic/store/synapse.py
Comment (BryanFauble) [inline schematic/store/synapse.py:1541]: In this file 2 sets of Annotations from SYNPY are being used:
```
from synapseclient.models.annotations import Annotations # This is the new set of Annotations that has Async support
from synapseclient import Annotations as OldAnnotations # This is the old set of Annotations that works with the current Entity objects
```
Since all of the Entities returned from `syn.get()` still use `OldAnnotations` I am using it in this context.

File: schematic/store/synapse.py
Comment (BryanFauble) [inline schematic/store/synapse.py:1607]: Doing this prevents the need to call `changeFileMetaData` as the file will be uploaded with the expected name. That function was/is super slow so this is a big gain.

File: schematic/store/synapse.py
Comment (BryanFauble) [inline schematic/store/synapse.py:1806]: I pulled in some changes from: https://github.com/Sage-Bionetworks/schematic/pull/1485 so I could make sure everything was working. I will handle any merge conflicts if that code goes first.

File: schematic/store/synapse.py
Comment (BryanFauble) [inline schematic/store/synapse.py:3074]: Moved to the non-deprecated function (Same behavior)

File: schematic_api/api/routes.py
Comment (BryanFauble) [inline schematic_api/api/routes.py:116]: This is no longer needed with the request hooks that I added for the `FlaskInstrumentor`

File: schematic_api/api/routes.py
Comment (BryanFauble) [inline schematic_api/api/routes.py:106]: This was failing with high concurrency as `tempfile.gettempdir()` only returned back `/tmp`, and for tests where the name of the uploaded file was the same was causing errors. By doing this we are guaranteed to get back a unique temporary directory.

File: schematic/store/database/README.md
Comment (thomasyu888) [inline schematic/store/database/README.md:1]: Thanks for adding this readme!

File: .github/workflows/api_test.yml
Comment (BryanFauble) [inline .github/workflows/api_test.yml:46]: I pulled the changes from https://github.com/Sage-Bionetworks/schematic/pull/1507 - As my build started to fail. I will pull in develop when this is merged in.

File: schematic/models/GE_Helpers.py
Comment (thomasyu888) [reply schematic/models/GE_Helpers.py:120]: Should this be a comment?

File: schematic/store/synapse_tracker.py
Comment (thomasyu888) [inline schematic/store/synapse_tracker.py:117]: Nit: There is a syn._GET_Paginated if that makes it easier

File: schematic_api/api/routes.py
Comment (thomasyu888) [reply schematic_api/api/routes.py:106]: I wonder if this impacts deployment, not sure if the /tmp directory is mounted in any way (probably not)

**Review by thomasyu888 — COMMENTED** [review]
I did a quick initial scan but I'll leave it to @BWMac / @jaymedina / @linglp / @andrewelamb / @GiaJordan for further comments before I go deeper into it.

I found your self review very helpful!

File: schematic/store/synapse_tracker.py
Comment (linglp) [inline schematic/store/synapse_tracker.py:50]: I found the logic here a bit difficult to follow without a comment. Could you add: "Returns None if `retrieve_if_not_present` is False and the entity is not found in the cache" as a comment?

File: schematic/store/synapse_tracker.py
Comment (linglp) [inline schematic/store/synapse_tracker.py:73]: Thanks for adding this @BryanFauble ! I am curious what do you envision the use case for this?

File: schematic/store/synapse.py
Comment (linglp) [inline schematic/store/synapse.py:1640]: I want to make sure that I understand the logic here correctly. Is the following correct? : 
1)  locate a file whose name matches existing_file_name
2) if the first step is None, then locate a file whose name matches file_name_new
3) if both are None, then local_tracked_file_instance is None

File: schematic/store/synapse.py
Comment (linglp) [inline schematic/store/synapse.py:2300]: I am seeing that you made similar changes in other parts of the code base by calling out the name of the parameter function. Is that a good practice?

File: schematic/utils/io_utils.py
Comment (BryanFauble) [reply schematic/utils/io_utils.py:58]: It's this code:

```
                if os.path.isfile(file_path) and os.path.getmtime(file_path) < (
                    time.time() - time_delta_seconds
                ):
```

And `3600` (1 hour) is being passed as `time_delta_seconds`

File: schematic/store/synapse_tracker.py
Comment (BryanFauble) [reply schematic/store/synapse_tracker.py:41]: Keeping both leaves the door open to use the function to retrieve metadata about an entity without the possibility of downloading that file. Consider these cases:

```
synapse_entity_tracker.get(
    synapse_id="syn123",
    syn=self.syn,
    download_file=False,
    retrieve_if_not_present=False,
)
```

```
synapse_entity_tracker.get(
    synapse_id="syn123",
    syn=self.syn,
    download_file=True,
    retrieve_if_not_present=True,
)
```

```
synapse_entity_tracker.get(
    synapse_id="syn123",
    syn=self.syn,
    download_file=False,
    retrieve_if_not_present=True,
)
```

The following case is the only one not valid as it would not download the file:

```
synapse_entity_tracker.get(
    synapse_id="syn123",
    syn=self.syn,
    download_file=True,
    retrieve_if_not_present=False,
)
```

File: schematic/store/synapse_tracker.py
Comment (BryanFauble) [reply schematic/store/synapse_tracker.py:50]: Updated docstring to read:
```
        """Retrieves a Synapse entity from the cache if it exists, otherwise downloads
        the entity from Synapse and adds it to the cache.

        Args:
            synapse_id: The Synapse ID of the entity to retrieve.
            syn: A Synapse object.
            download_file: If True, download the file.
            retrieve_if_not_present: If True, retrieve the entity if it is not present 
                in the cache. If not found in the cache, and this is False, return None.
            download_location: The location to download the file to.
            if_collision: The action to take if there is a collision when downloading 
                the file. May be "overwrite.local", "keep.local", or "keep.both". A 
                collision occurs when a file with the same name already exists at the 
                download location.

        Returns:
            The Synapse entity if found. When retrieve_if_not_present is False and the
            entity is not found in the local cache, returns None. If
            retrieve_if_not_present is True and the entity is not found in the local
            cache, retrieve the entity from Synapse and add it to the cache.
        """
```

File: schematic/store/synapse_tracker.py
Comment (BryanFauble) [reply schematic/store/synapse_tracker.py:73]: See https://sagebionetworks.jira.com/browse/PLFM-8605 - We must remove entities from the local cache when the table data is updated.

File: schematic/store/synapse.py
Comment (BryanFauble) [reply schematic/store/synapse.py:1640]: Yes that's correct

File: schematic/store/synapse.py
Comment (BryanFauble) [reply schematic/store/synapse.py:2300]: Yes. IMO Positional arguments are only good when there are 1, maybe 2 arguments you can pass into a function. Any more then it makes it significantly more difficult to skim the code and understand what is going on.

File: schematic/store/synapse_tracker.py
Comment (BryanFauble) [reply schematic/store/synapse_tracker.py:117]: That, nor `_POST_paginated` would appear to work in this case as `nextPageToken` is a query parameter, instead of a piece of the body.

The return type of:
https://rest-docs.synapse.org/rest/GET/projects/user/principalId.html

Is different in it's format to the `_GET_Paginated` expected type of:
https://rest-docs.synapse.org/rest/GET/teamMembers/id.html

File: schematic/models/GE_Helpers.py
Comment (BryanFauble) [reply schematic/models/GE_Helpers.py:365]: Gotcha @GiaJordan What are your thoughts if I only call `self.context.build_data_docs(resource_identifiers=[suite_identifier])` if requested by the person running the code? This means when schematic is running as an API server it will never build these HTML files.

This is extra processing currently occurring in production schematic that is not needed. It's small, but non-zero.

On a side note: Do we have any environment variables/flags I could re-use that area already used in other areas of the code?

File: schematic/manifest/commands.py
Comment (BryanFauble) [reply schematic/manifest/commands.py:350]: Added! @linglp

File: schematic/models/GE_Helpers.py
Comment (BryanFauble) [reply schematic/models/GE_Helpers.py:118]: Added!

File: schematic/models/GE_Helpers.py
Comment (GiaJordan) [reply schematic/models/GE_Helpers.py:363]: @BryanFauble yeah that's fine. And there isn't one we've made that can be reused, but I think it would make sense to tie it to the verbosity requested for the command and include both lines like
```
if logger.isEnabledFor(logging.DEBUG):
    self.context.build_data_docs(resource_identifiers=[suite_identifier])
    #Webpage DataDocs opened here:
    self.context.open_data_docs(resource_identifier=suite_identifier)
```
It won't log to the logger, but `open_data_docs` will open the html file with the useful information in a browser and won't interrupt the rest of the run

File: schematic/store/synapse_tracker.py
Comment (BryanFauble) [reply schematic/store/synapse_tracker.py]: Added!

File: schematic/store/synapse_tracker.py
Comment (BryanFauble) [reply schematic/store/synapse_tracker.py]: Added!

File: schematic/models/GE_Helpers.py
Comment (BryanFauble) [reply schematic/models/GE_Helpers.py]: Perfect @GiaJordan I updated this code to:

```

        if logger.isEnabledFor(logging.DEBUG):
            self.context.build_data_docs(resource_identifiers=[suite_identifier])
            ##Webpage DataDocs opened here:
            # self.context.open_data_docs(resource_identifier=suite_identifier)
```

I am going to leave `open_data_docs` commented out like it was.

File: schematic/store/synapse.py
Comment (linglp) [reply schematic/store/synapse.py:2300]: Got it Bryan! Thanks for the explanation.

File: schematic/store/synapse.py
Comment (linglp) [reply schematic/store/synapse.py:1640]: Thanks Bryan!

File: schematic/manifest/commands.py
Comment (linglp) [reply schematic/manifest/commands.py:350]: Thanks Bryan!

File: schematic/store/synapse_tracker.py
Comment (linglp) [reply schematic/store/synapse_tracker.py:80]: Hi @BryanFauble ! Thanks for providing more context. I read the Jira ticket that you linked but I don't think that I understand. Based on the ticket and the comment, I am guessing that `store` function is not returning the etag of table because UploadToTableResult.etag is not the etag of the table. But why is it related to removing entities from the local cache?

Comment (BryanFauble) [reply schematic/store/synapse_tracker.py:80]: @linglp Let's recap the concept of eTag and what it's used for in Synapse:

> Synapse employs an Optimistic Concurrency Control (OCC) scheme to handle concurrent updates. Since the E-Tag changes every time an entity is updated it is used to detect when a client's current representation of an entity is out-of-date.

We can take a simple example where suppose we create a Table and it's assigned an etag of `1`, I go ahead and add an annotation to the table and that Table now has a new eTag `2`. When I updated those annotations we will want to make sure that we capture in the local cache:

1. The current state of the entity (This includes all of the traditional fields about the Entity, AND, the annotations for that entity)
2. The updated eTag as a result of saving those annotations.

The above works because when we make the change to add those annotations the Synapse API gives us back the eTag of the updated entity and we can store that back into our local in-memory cache.


In the case of updates to the data or schema of a table the eTag is also getting updated. Let's suppose we add data to a table, Synapse will now update that eTag to `3`, however, us as the caller have no way of knowing that Synapse updated the eTag to `3`, we only ever know it was `2`, and we never got the updated eTag as a result of submitting our transaction. 

Ask: What would occur if I try to submit another transaction on that Table and pass along my known eTag of `2`? Synapse would reject the transaction because we on the client side do not have the most up to date version of that entity that Synapse has. This is the root of what the ticket is talking about/addressing.


Why do we need to remove the entity from our local in-memory cache? It's the result of the `Ask` above. Since we do not have the most up to date eTag of that entity, the next time we read for that entity in order to perform some operation, that read MUST be sent to Synapse. We cannot trust our local in-memory cache anymore because we know we do not have the up to date eTag of the entity. That is what the `.remove()` operation is for.

File: schematic/store/synapse.py
Comment (andrewelamb) [inline schematic/store/synapse.py:91]: Entity has a typo

File: schematic/store/synapse.py
Comment (andrewelamb) [inline schematic/store/synapse.py:176]: I see some use of protected attibutes in the File Enntity class. Are these somethign that could be made pubic in the Python client in the near future?

File: schematic/store/synapse.py
Comment (andrewelamb) [reply schematic/store/synapse.py:176]: I'm curious why renaming the file is a concern. Is that something schematic was doing?

File: schematic/store/synapse_tracker.py
Comment (andrewelamb) [reply schematic/store/synapse_tracker.py:1]: This question is probably due to how little time I ahve spent in this module. I'm curious why this was needed in the first place? Was Schematic repeatedly downloading the same entity in the same API call?

File: schematic/store/synapse.py
Comment (BryanFauble) [reply schematic/store/synapse.py:176]: The intention is that these FileEntities are going to be deprecated/removed to be replaced by this interface:
https://python-docs.synapse.org/reference/oop/models_async/#synapseclient.models.File

File: schematic/store/synapse.py
Comment (BryanFauble) [reply schematic/store/synapse.py:91]: lol. I can't believe I hadn't noticed this before :D

Patched!

File: schematic/store/synapse_tracker.py
Comment (BryanFauble) [reply schematic/store/synapse_tracker.py:1]: > Was Schematic repeatedly downloading the same entity in the same API call?

Yes. Early in the script metadata about an entity was retrieved from Synapse. We can take the manifest FileEntity for example. Every time that manifest needed to be referenced the metadata from Synapse was re-retrieved.

In the benchmarking that I had done it cut down a significant number of calls to Synapse. In the screenshot I ran all tests prefixed with `test_submit_%` and the total number of HTTP calls went from 972 -> 594.
image

**Review by andrewelamb — APPROVED** [review]

**Pull request merged**
